sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InsecureDeserialization.cs

```diff
@@ -0,0 +1,283 @@
+
+                var hasConditionalStatements = isDeserializationCtor
+                    ? node.ContainsConditionalConstructs()
+                    : HasParametersUsedInConditionalConstructs(node);
+
+                constructorsInfo.Add(new ConstructorInfo(node, hasConditionalStatements, isDeserializationCtor));
+
+                base.VisitConstructorDeclaration(node);
+            }
+
+            public override void VisitClassDeclaration(ClassDeclarationSyntax node)
+            {
+                if (classLevel != 0)
+                {
+                    // Skip nested class visits. The rule will be triggered for them also.
+                    return;
+                }
+
+                classLevel++;
+                base.VisitClassDeclaration(node);
+            }
+
+            private bool HasParametersUsedInConditionalConstructs(BaseMethodDeclarationSyntax declaration)
+            {
+                var symbols = GetConstructorParameterSymbols(declaration, semanticModel);
+
+                var conditionalsWalker = new ConditionalsWalker(semanticModel, symbols);
+                conditionalsWalker.SafeVisit(declaration);
+
+                return conditionalsWalker.HasParametersUsedInConditionalConstructs;
+            }
+
+            private ConstructorInfo GetDeserializationConstructor() =>
+                constructorsInfo.SingleOrDefault(info => info.IsDeserializationConstructor);
+
+            private bool IsDeserializationConstructor(BaseMethodDeclarationSyntax declaration) =>
+                // A deserialization ctor has the following parameters: (SerializationInfo information, StreamingContext context)
+                // See https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.iserializable?view=netcore-3.1#remarks
+                declaration.ParameterList.Parameters.Count == 2 &&
+                declaration.ParameterList.Parameters[0].IsDeclarationKnownType(KnownType.System_Runtime_Serialization_SerializationInfo, semanticModel) &&
+                declaration.ParameterList.Parameters[1].IsDeclarationKnownType(KnownType.System_Runtime_Serialization_StreamingContext, semanticModel);
+
+            private static ImmutableArray<ISymbol> GetConstructorParameterSymbols(BaseMethodDeclarationSyntax node, SemanticModel semanticModel) =>
+                node.ParameterList.Parameters
+                    .Select(syntax => GetParameterSymbols(syntax, semanticModel))
+                    .ToImmutableArray();
+
+            private static ISymbol GetParameterSymbols(SyntaxNode parameterSyntax, SemanticModel semanticModel) =>
+                semanticModel.GetDeclaredSymbol(parameterSyntax);
+        }
+
+        /// <summary>
+        /// This walker is responsible to visit all conditional structures and check if a list of parameters
+        /// are used or not.
+        /// </summary>
+        private sealed class ConditionalsWalker : CSharpSyntaxWalker
```

Comment:
Question regarding switch statement support also relates to `ConditionalsWalker`